doc/api_result.rst

```diff
@@ -0,0 +1,177 @@
+* various information to help post-mortem investigation, should any problem
+  occurred (logs, list of subprocesses spawned during test execution,
+  environment variables, ...).
+
+Even though test drivers create a default ``TestResult`` instance
+(``self.result`` test driver attribute), the actual registration of test
+results to the testsuite report is manual: test drivers must call their
+``push_result`` method for that. This is how a single test driver instance
+(i.e. a single testcase) can register multiple test results.
+
+The typical use case for a single driver instance pushing multiple test results
+is for testcases that contain multiple "inputs" and 1) compile a test program
+2) run that program once for each input. In this case, it makes sense to create
+one test result per input, describing whether the software behaves as expected
+for each one independently rather than creating a single result that describes
+whether the software behaved well for *all* inputs.
+
+This leads to the role of the test result name (``test_name`` attribute of
+``TestResult`` instances). The name of the default result that drivers create
+is simply the name of the testcase. This is a nice start, since it makes it
+super easy for someone looking at the report to relate the ``FAIL foo-bar``
+result to the ``foo/bar`` testcase. By convention, drivers that create multiple
+results assign them names such as ``TEST_NAME.INPUT_NAME``, i.e. just put a dot
+between the testcase name and the name of the input that triggers the emission
+of a separate result.
+
+An example may help to clarify. Imagine a testsuite for a JSON handling
+library, and the following testcase that builds a test program that 1) parses
+its JSON input 2) pretty-prints that JSON document on its output:
+
+.. code-block:: text
+
+   parsing/
+      test.yaml
+      test_program.c
+
+      empty-array.json
+      empty-array.out
+
+      zero.json
+      zero.out
+
+      ...
+
+A test driver for this testcase could do the following:
+
+* Build ``test_program.c``, a program using the library to test (no test result
+  for that).
+* Run that program on ``empty-array.json``, compare its output to
+  ``empty-array.out`` and create a test result for that comparison, whose name
+  is ``parsing.empty-array``.
+* Likewise for ``zero.json`` and ``zero.out``, creating a ``parsing.zero`` test
+  result.
+* ...
+
+
+.. _api_test_status:
+
+``TestStatus``
+--------------
+
+This is an ``Enum`` subclass, allowing to classify results: tests that passed
+(``TestStatus.PASS``), tests that failed (``TestStatus.FAIL``), etc. For
+convenience, here the list of all available statuses as described in the
+``result.py`` module:
+
+PASS
+   The test has run to completion and has succeeded.
+
+FAIL
+   The test has run enough for the testsuite to consider that it failed.
+
+XFAIL
+   The test has run enough for the testsuite to consider that it failed, and
+   that this failure was expected.
+
+XPASS
```

Comment:
On one hand, I didn’t want to do that because this is AdaCore-centric legacy info whereas, except for the “adacore” section, this whole doc is supposed to be standalone and readable by outsiders (this is a public project), but on the other hand I assume this can be useful and the info is short enough for this violation of principle to be minor. So I guess I’ll just add the info. Thanks.